### ISSUE TITLE
focusgroup attribute

### ISSUE THREAD

**Issue by lukewarlow** [WebKit/standards-positions#171]
### Title of the spec

focusgroup attribute

### URL to the spec

https://github.com/whatwg/html/pull/11723

### URL to the spec's repository

https://github.com/whatwg/html/

### Explainer URL

https://open-ui.org/components/scoped-focusgroup.explainer/

### TAG Design Review URL

https://github.com/w3ctag/design-reviews/issues/1152

### Mozilla standards-positions issue URL

https://github.com/mozilla/standards-positions/issues/631

### WebKit Bugzilla URL

https://bugs.webkit.org/show_bug.cgi?id=255482

### Radar URL

rdar://108403002

### Description

A new HTML attribute that allows for creating focusgroups that are navigable using arrow keys without needing to use JavaScript. (See explainer for a far better explanation)

**Comment by hober** [comment]
Added `concerns: venue` because this should be discussed in the HTML Workstream at WHATWG.

**Comment by cookiecrook** [comment]
Overall I think the `focusgroup` proposal is a good idea, will be useful to authors, and will be more accessible to users. I'd like to see an example 

Stated under "Non-Goals" heading:
> For this proposed standard, we decouple selection from focus and only focus (cough) on the focus aspect: moving focus using the keyboard among eligible focus targets. Adding selection state, e.g., tracking the currently selected option in a list, will be a separate feature. To handle selection-based scenarios, we defer to the CSS proposal [Toggles](https://tabatkins.github.io/css-toggle/) which can work nicely with focusgroups.

This intended decoupling of focus and selection may require AT changes (e.g. VoiceOver) in addition to User Agent (WebKit) additions. As such, I think we would need to see an example of the Toggles+FocusGroup combo example that is alluded to here. Especially for an interactive ARIA grid, with multiple selection and multi-directional focus via `focus-group-direction: horizontal;` perhaps including the wrap options.

**Comment by cookiecrook** [comment]
Another item I think is missing from the proposal is that an extension interface or assistive technology (e.g. VoiceOver or another) should be able to drive this "roaming/roving tabindex" selection if needed. For example, by calling UA focus() on one of the tabindex="-1" elements, in the same way that a mouse user could click to set focus directly, even if it's not Tab-focusable.

**Comment by lukewarlow** [comment]
Any concerns (aside from the venue itself) or additions I recommend filing issues for in the open-ui GitHub.

**Comment by cookiecrook** [comment]
Assuming there's a missing question mark after the word additions above. In any case, I'm having trouble parsing... Also that the venue concern appears acknowledged but then dismissed in the same sentence.

**Comment by lukewarlow** [comment]
I'm just a web developer, not someone working on this proposal. 

I merely meant any feedback you have would be welcomed in the open UI repo so they can integrate that feedback into the API design.

I've raised an issue to query moving the spec to a WHATWG/html PR but I suspect this would require multi browser interest to champion and land this API.

**Comment by janewman** [comment]
Hi folks, the focusgroup explainer has evolved since this was last visited, and I recently created a new, "Scoped" version of that explainer that takes all of these advancements and then scopes the behavior in a few different ways:
* Moves 2D grid navigation to a future consideration.
* Moves CSS support to a future consideration.
* Requires a new "behavior" attribute that will be used to compute a minimum aria role, ensuring that focusgroups are accessible-by-default.

I would love to get feedback on the approach; I intended for this to resolve the biggest remaining points of contention.
https://open-ui.org/components/scoped-focusgroup.explainer/

**Comment by lukewarlow** [comment]
Removing venue concerns because it's now being progressed through the WHATWG stages.

PR: https://github.com/whatwg/html/pull/11723
Issue: https://github.com/whatwg/html/issues/11641

**Comment by annevk** [comment]
Apologies for the long delay here. We like the general idea, but we have some concerns about the specific user interaction model.

We don't expect that all platforms will use arrow keys to navigate between the focusable elements. That should be abstracted, also at the test level. And the tests assume that keyboard navigation is enabled, which it isn't by default on macOS. That should also have some kind of WebDriver toggle that is standardized and the tests use.

**Comment by janewman** [comment]
> Apologies for the long delay here. We like the general idea, but we have some concerns about the specific user interaction model.
> 
> We don't expect that all platforms will use arrow keys to navigate between the focusable elements. That should be abstracted, also at the test level. And the tests assume that keyboard navigation is enabled, which it isn't by default on macOS. That should also have some kind of WebDriver toggle that is standardized and the tests use.

Appreciate this feedback, and I couldn't be more aligned that this feature should not be limited to standard keyboard arrow keys. I've refactored the tentative tests to abstract away this except in a utility. Ideally, I'd like this utility to take in a preference given by the user agent, but I'll need to spend some more time here to see what our options are to have user agents specify such a preference to WebDriver. Regardless, this would not result in an API change.

As for respecting the keyboard navigation setting in MacOS, it doesn't look to me like WebDriver currently has the ability to toggle MacOS settings, in the meantime I've adjusted all tests to be agnostic of this setting by using an author-specified tab index of 0.
